crates/noirc_frontend/src/parser/parser.rs

```diff
@@ -235,12 +236,22 @@ fn lambda_return_type() -> impl NoirParser<UnresolvedType> {
         .map(|ret| ret.unwrap_or(UnresolvedType::Unspecified))
 }
 
-fn function_return_type() -> impl NoirParser<(AbiVisibility, UnresolvedType)> {
+fn function_return_type() -> impl NoirParser<(AbiVisibility, AbiDistinctness, UnresolvedType)> {
     just(Token::Arrow)
         .ignore_then(optional_visibility())
+        .then(optional_distinctness())
```

Comment:
Note for future documentation: parsing it this way means we will only parse it if `distinct` comes after `pub`. That is, `fn foo() -> pub distinct Field { 0 }` is valid and `fn foo() -> distinct pub Field { 0 }` is a parse error.

Reply:
Thats a good point -- Comparing the difference, `distinct pub Field` reads a bit more clearer. Having it be in one position seems fine to me

Reply:
I'll swap them round